File: 189-rotate-array/rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        k=k%len(nums)
        nums[::]=nums[-k:]+nums[:-k]
        """

        def rev(nums,left,right):
            while left<=right:
                nums[left],nums[right]=nums[right],nums[left]
                left+=1
                right-=1
        #-------------------------------------------------

        k=k%len(nums)
        n=len(nums)-1
        rev(nums,0,n)
        rev(nums,0,k-1)
        rev(nums,k,n)
        

File: 189-rotate-array/rotate-array.java
class Solution {
    public void reverse(int[] nums,int left,int right){

        while(left<right){
            int temp=nums[left];
            nums[left]=nums[right];
            nums[right]=temp;

            left++;
            right--;
        }

    }
    public void rotate(int[] nums, int k) {

        int n=nums.length;
        k=k%n;

        reverse(nums,0,n-1); 
        reverse(nums,0,k-1); 
        reverse(nums,k,n-1); 
    }
}


/*
int n=nums.length;
        k=k%n; //maximum length no of rotations

        for(int j=1;j<=k;j++){

            int last=nums[n-1];
            for(int i=n-1;i>=1;i--){
                nums[i]=nums[i-1];
            }
            nums[0]=last;

        }

*/